app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -343,4 +352,55 @@ public void onRequestPermissionsResult(int requestCode, @NonNull String[] permis
         }
     }
 
+    /**
+     * Perform secure key generation in an AsyncTask.
+     */
+    private void onKeyGenerationSlideShown() {
+        mNextButton.setVisibility(View.GONE);
+        KeyGenerationTask keyGenerationTask = new KeyGenerationTask(this);
+        keyGenerationTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+    }
+
+    /**
+     * Sets up the initial configuration and generates secure keys.
+     */
+    private static class KeyGenerationTask extends AsyncTask<Void, Void, Void> {
+        private WeakReference<FirstStartActivity> refFirstStartActivity;
+        ConfigXml configXml;
+
+        KeyGenerationTask(FirstStartActivity context) {
+            refFirstStartActivity = new WeakReference<>(context);
+        }
+
+        @Override
+        protected Void doInBackground(Void... voids) {
+            FirstStartActivity firstStartActivity = refFirstStartActivity.get();
+            if (firstStartActivity == null) {
+                cancel(true);
+                return null;
+            }
```

Comment:
Is this the condition whether this slide is shown or not?  
I am asking because naively I would have expected a condition when setting the layout, as I would expect that's relevant there (e.g. for the dots at the bottom indicating the current and amount of slides). Remember, I know nothing about this stuff, so this isn't criticism.

Reply:
No, this is the condition when the AsyncTask ("a separate thread") should end immediately. If this occurs (rarely) something has gone particularly wrong because the activity is destroyed before the asyncTask completed its work. As I implemented it like "don't allow the user to advance out of this activity" this should not happen. But in case another one changes code around, this is a very precautionary approach to be better safe than sorry. Best practice in Android coding say, you should check if your activity (the caller) is still there before referencing it from an AsyncTask.

Reply:
( or what's preventing you from doing that? )
I'm not sure if will work okay as I need this exact class FirstStartActivity again for my next improvement. I often experienced that a lot of minor changes are necessary and then I don't know how to keep the branch-branch in sync properly. (I did then create a new branch from the master and manually reapplied the changed code parts). Can I also pull from merged-branch to sub-branch when the point where the sub-branch was created isn't the commit that got merged? (and having less conflicts to resolve)
I just meant it would be a nice bonus if things go faster, but it's also okay if we stick to a good review that takes a little bit longer.

To your question above:
The app currently does only check if a crucial permission is missing (like storage permission) and if so repeat the welcome assistant. I think there are pros and cons to this.
Pro: Code is shorter and less error-prone as we don't have to reorder slides an have "hardcoded int's" versus the "we can't get a hardcoded one, so a constant is required in code"
Cons: More code, more checks and reorder of the slides. The simple approach checking for "ready for the next slide" wouldn't apply anymore. We also would required saving bundles for the slide states for the case the phone is rotated > much more work. If necessary, I can do it.

... So long no one complained having to go through all slides again and that goes real quick. If something is already done you can press Continue>Continue>Continue and land. Repeating the slides also improves migration tasks if we later need to change something about the storage permission (I made this up, but can happen). We would then have overhead for checking if we need it and if we need a migration step to show the slide again.

Reply:
So showing pointless slides is very confusing to the user. I imagine users screaming on the forum now wanting to loose their device ID after revoking storage permissions or something like that (as slides imply the keys will be generated).

 You could simply build a list of strings indicating what needs to be shown, and just serialize that list between rotations.